operators/multiclusterobservability/controllers/multiclusterobservability/analytics/rightsizing_namespace_policy.go

```diff
@@ -0,0 +1,192 @@
+
+	"github.com/cloudflare/cfssl/log"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	configpolicyv1 "open-cluster-management.io/config-policy-controller/api/v1"
+	policyv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func modifyComplianceTypeIfPolicyExists(c client.Client) error {
+	policy := &policyv1.Policy{}
+	err := c.Get(context.TODO(), types.NamespacedName{
+		Name:      rsPrometheusRulePolicyName,
+		Namespace: rsNamespace,
+	}, policy)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Policy does not exist. Skipping update.")
+			return nil
+		}
+		log.Error(err, "Error retrieving the policy")
+		return err
+	}
+
+	// Unmarshal the inner ConfigurationPolicy
+	for _, pt := range policy.Spec.PolicyTemplates {
+		var configPolicy configpolicyv1.ConfigurationPolicy
+		err := json.Unmarshal(pt.ObjectDefinition.Raw, &configPolicy)
+		if err != nil {
+			log.Error(err, "Failed to unmarshal ConfigurationPolicy from PolicyTemplate")
+			return err
+		}
+
+		// Change ComplianceType if it's "MustOnlyHave"
+		changed := false
+		for _, objTemplate := range configPolicy.Spec.ObjectTemplates {
+			if objTemplate.ComplianceType == "MustOnlyHave" {
+				objTemplate.ComplianceType = "MustNotHave"
+				changed = true
+			}
+		}
+
+		if changed {
+			// Marshal the modified ConfigurationPolicy back into JSON
+			modifiedRaw, err := json.Marshal(configPolicy)
+			if err != nil {
+				log.Error(err, "Failed to marshal modified ConfigurationPolicy")
+				return err
+			}
+
+			pt.ObjectDefinition = runtime.RawExtension{Raw: modifiedRaw}
+		}
+	}
+
+	// Update the modified policy
+	err = c.Update(context.TODO(), policy)
+	if err != nil {
+		log.Error(err, "Failed to update the modified policy")
+		return err
+	}
+
+	log.Info("Successfully updated ComplianceType in policy")
+
+	// Wait for 1 seconds
+	time.Sleep(1 * time.Second)
+
+	return nil
+}
+
+func createOrUpdatePrometheusRulePolicy(c client.Client, prometheusRule monitoringv1.PrometheusRule) error {
+
+	policy := &policyv1.Policy{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "policy.open-cluster-management.io/v1",
+			Kind:       "Policy",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rsPrometheusRulePolicyName,
+			Namespace: rsNamespace,
+		},
+	}
+	log.Info("RS - Policy object created")
+
+	errPolicy := c.Get(context.TODO(), types.NamespacedName{
+		Name:      policy.Name,
+		Namespace: policy.Namespace,
+	}, policy)
+
+	if errPolicy != nil && !errors.IsNotFound(errPolicy) {
+		fmt.Println("Error retriving Policy:", errPolicy)
+		return errPolicy
+	}
+
+	// Marshal the PrometheusRule object into JSON
+	promRuleJSON, err := addAPIVersionAndKind(prometheusRule, "monitoring.coreos.com/v1", "PrometheusRule")
```

Comment:
Maybe we could use `runtime.Convert_runtime_Object_To_runtime_RawExtension` instead? :thinking:

Reply:
Good idea, implemented the code as suggested, which avoids the need for JSON marshaling/unmarshaling..